pkg/cloudprovider/servicecontroller/servicecontroller.go

```diff
@@ -0,0 +1,407 @@
+	}
+
+	if publicIPstring == fmt.Sprint(service.Spec.PublicIPs) {
+		glog.Infof("Not persisting unchanged service to registry.")
+		return nil
+	}
+	s.setService(namespacedName, service)
+
+	// If creating the load balancer succeeded, persist the updated service.
+	if err = s.persistUpdate(service); err != nil {
+		glog.Errorf("Failed to persist updated publicIPs to apiserver, even after retries. Giving up: %v", err)
+		return nil
+	}
+	return nil
+}
+
+// TODO(a-robinson): Handle repeated failures due to ResourceVersion changes or
+// the object having been deleted.
+func (s *ServiceController) persistUpdate(service *api.Service) error {
+	var err error
+	for i := 0; i < clientRetryCount; i++ {
+		_, err = s.kubeClient.Services(service.Namespace).Update(service)
+		if err == nil {
+			return nil
+		}
+		glog.Warningf("Failed to persist updated PublicIPs to service %s after creating its external load balancer: %v",
+			service.Name, err)
+		time.Sleep(clientRetryInterval)
+	}
+	return err
+}
+
+func (s *ServiceController) createExternalLoadBalancer(service *api.Service) error {
+	ports, err := getTCPPorts(service)
+	if err != nil {
+		return err
+	}
+	nodes, err := s.kubeClient.Nodes().List()
+	if err != nil {
+		return err
+	}
+	name := s.loadBalancerName(service)
+	if len(service.Spec.PublicIPs) > 0 {
+		for _, publicIP := range service.Spec.PublicIPs {
+			// TODO: Make this actually work for multiple IPs by using different
+			// names for each. For now, we'll just create the first and break.
+			endpoint, err := s.balancer.CreateTCPLoadBalancer(name, s.zone.Region, net.ParseIP(publicIP),
+				ports, hostsFromNodeList(nodes), service.Spec.SessionAffinity)
+			if err != nil {
+				return err
+			}
+			service.Spec.PublicIPs = []string{endpoint}
+			break
+		}
+	} else {
+		endpoint, err := s.balancer.CreateTCPLoadBalancer(name, s.zone.Region, nil,
+			ports, hostsFromNodeList(nodes), service.Spec.SessionAffinity)
+		if err != nil {
+			return err
+		}
+		service.Spec.PublicIPs = []string{endpoint}
+	}
+	return nil
+}
+
+func (s *ServiceController) handleDelete(service *api.Service) error {
+	if err := s.deleteExternalLoadBalancer(service); err != nil {
+		return err
+	}
+	namespacedName, err := cache.MetaNamespaceKeyFunc(service)
+	if err != nil {
+		// This is panic-worthy, since the queue shouldn't have been able to
+		// handle the service if it couldn't generate a name for it.
+		glog.Errorf("Couldn't generate namespaced name for service: %v", err)
+		return nil
+	}
+	s.deleteService(namespacedName)
+	return nil
+}
+
+func (s *ServiceController) deleteExternalLoadBalancer(service *api.Service) error {
+	if err := s.balancer.DeleteTCPLoadBalancer(s.loadBalancerName(service), s.zone.Region); err != nil {
+		return err
+	}
+	return nil
+}
+
+// ListKeys implements the interface required by DeltaFIFO to list the keys we
+// already know about.
+func (s *ServiceController) ListKeys() []string {
```

Comment:
I hadn't noticed the KeyListerFunc type initially, that should do the trick for wrapping this in a closure.